### PR TITLE
Added CRUD operations for managing roles in datastore

### PR DESCRIPTION
**Pull request by poadhika**
Added CRUD operations for storing roles in datastore
Allows Add/Update/Delete  for single roles
Returns single or all roles based on query

Addresses issue: https://github.com/Microsoft/fhir-server/issues/278

Tests needs to be added

### CODE REVIEW

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 1]
```diff
@@ -49,6 +51,8 @@ public class ControlPlaneDataStore : IControlPlaneDataStore
 
             _documentClient = documentClient;
             _collectionUri = cosmosDataStoreConfiguration.GetRelativeCollectionUri(collectionConfig.CollectionId);
+            _collectionId = cosmosDataStoreConfiguration.ControlPlaneCollectionId;
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: Should use `collectionConfig.CollectionId` instead.

File: src/Microsoft.Health.CosmosDb/Configs/CosmosDataStoreConfiguration.cs [hunk 2]
```diff
@@ -20,6 +20,14 @@ public class CosmosDataStoreConfiguration
 
         public int? InitialDatabaseThroughput { get; set; }
 
+        public string FhirCollectionId { get; set; }
```
Comment (brandonpollett) [inline src/Microsoft.Health.CosmosDb/Configs/CosmosDataStoreConfiguration.cs]: These properties shouldn't be added. They were refactored out to the various collection configs.

File: src/Microsoft.Health.Fhir.Api/Controllers/ControlPlaneController.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            var response = await _rbacService.GetIdentityProviderAsync(identityProviderName, cancellationToken);
+            return Ok(response);
+        }
+
+        [HttpGet]
+        [Route("Admin/Roles/{RoleName}")]
+        [AllowAnonymous]
+        public async Task<IActionResult> GetRole(string roleName, CancellationToken cancellationToken)
+        {
+            if (roleName != "all")
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Api/Controllers/ControlPlaneController.cs]: I would move this such that a `GET` on `/Admin/Roles/` returns all

File: test/Microsoft.Health.Fhir.Tests.E2E/Properties/launchSettings.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (brandonpollett) [inline test/Microsoft.Health.Fhir.Tests.E2E/Properties/launchSettings.json]: Was this file intentionally added?

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCollectionInitializer.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿// -------------------------------------------------------------------------------------------------
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCollectionInitializer.cs]: this file shouldn't exist anymore.

File: src/Microsoft.Health.Fhir.Api/Controllers/ControlPlaneController.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
+﻿// -------------------------------------------------------------------------------------------------
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Api/Controllers/ControlPlaneController.cs]: Let's hold off on adding this yet since we don't have anyway to lock it down.

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Rbac/CosmosRole.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        {
+        }
+
+        [JsonProperty("id")]
+        public string Id => Name;
+
+        [JsonProperty("partitionKey")]
+        public string PartitionKey { get; } = RolePartition;
+
+        [JsonProperty("_etag")]
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Rbac/CosmosRole.cs]: There's a `KnownDocumentProperties.Etag` now. Also, there are known ones for `PartitionKey` and `Id`

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Rbac/CosmosRole.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+using Microsoft.Health.ControlPlane.Core.Features.Rbac.Roles;
+using Microsoft.Health.CosmosDb.Features.Storage;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.ControlPlane.CosmosDb.Features.Storage.Rbac
+{
+    public class CosmosRole : Role
+    {
+        internal const string RolePartition = "Role_";
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Rbac/CosmosRole.cs]: convention should be `_roles`

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Rbac/CosmosRole.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+        public CosmosRole(Role role)
+              : base(role.Name, role.ResourcePermissions,  role.Version)
+        {
+            ETag = $"\"{role.Version}\"";
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/Rbac/CosmosRole.cs]: We probably don't actually need this since the version will always be set by CosmosDB.

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 10]
```diff
@@ -66,13 +70,70 @@ public async Task<IdentityProvider> GetIdentityProviderAsync(string name, Cancel
...
+        public async Task<Role> UpsertRoleAsync(Role role, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(role, nameof(role));
+
+            var cosmosRole = new CosmosRole(role);
+            var resultRole = await UpsertSystemObjectAsync(cosmosRole, CosmosRole.RolePartition, cancellationToken);
+            return resultRole.ToRole();
+        }
+
+        public async Task<Role> AddRoleAsync(Role role, CancellationToken cancellationToken)
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: Add isn't needed, it can be achieved with `UpsertRoleAsync`

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 11]
```diff
@@ -118,6 +176,32 @@ private async Task<T> GetSystemDocumentByIdAsync<T>(string id, string partitionK
             }
         }
 
+        private async Task<IEnumerable<T>> GetSystemDocumentAllAsync<T>(CancellationToken cancellationToken)
+        {
+            var documentQuery = new SqlQuerySpec(
+                "SELECT * FROM root");
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: This is going to get everything in the database. We should be able to make it a partitioned query and only return items in the `_roles` partition to get everything.

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 12]
```diff
@@ -66,13 +70,70 @@ public async Task<IdentityProvider> GetIdentityProviderAsync(string name, Cancel
...
+            }
+
+            return response;
+        }
+
+        public async Task<IEnumerable<Role>> GetRoleAllAsync(CancellationToken cancellationToken)
+        {
+            var role = await GetSystemDocumentAllAsync<CosmosRole>(cancellationToken);
+
+            if (role == null)
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: Not having any roles in the system isn't necessarily an error.

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 13]
```diff
@@ -129,26 +213,75 @@ private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitio
...
                     true,
                     cancellationToken);
                 _logger.LogInformation("Request charge: {RequestCharge}, latency: {RequestLatency}", response.RequestCharge, response.RequestLatency);
-
-                return (T)(dynamic)response.Resource;
+                return (dynamic)response.Resource;
             }
             catch (DocumentClientException dce)
             {
-                if (dce.StatusCode == HttpStatusCode.TooManyRequests)
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: Looks like the check for `HttpStatusCode.TooManyRequests` was accidentally removed. Also need to check this in the other queries, too.

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 14]
```diff
@@ -129,26 +213,75 @@ private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitio
...
+                ResourceResponse<Document> response = await _documentClient.Value.CreateDocumentAsync(
                     _collectionUri,
                     systemObject,
                     requestOptions,
                     true,
                     cancellationToken);
                 _logger.LogInformation("Request charge: {RequestCharge}, latency: {RequestLatency}", response.RequestCharge, response.RequestLatency);
-
-                return (T)(dynamic)response.Resource;
+                return (dynamic)response.Resource;
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: Was this intentionally changed from `return (T)(dynamic)response.Resource`?

File: src/Microsoft.Health.CosmosDb/Configs/CosmosDataStoreConfiguration.cs [hunk 15]
```diff
@@ -20,6 +20,12 @@ public class CosmosDataStoreConfiguration
 
         public int? InitialDatabaseThroughput { get; set; }
 
+        public int? InitialFhirCollectionThroughput { get; set; }
```
Comment (brandonpollett) [inline src/Microsoft.Health.CosmosDb/Configs/CosmosDataStoreConfiguration.cs]: The `InitialFhirCollectionThroughput`, `ControlPlaneCollectionId`, and `InitialControlPlaneCollectionThroughput` aren't needed either.

File: src/Microsoft.Health.ControlPlane.CosmosDb/Microsoft.Health.ControlPlane.CosmosDb.csproj [hunk 16]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <CodeAnalysisRuleSet>..\..\CustomAnalysisRules.ruleset</CodeAnalysisRuleSet>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Microsoft.Health.ControlPlane.CosmosDb.csproj]: Was this change intentional?

File: test/Microsoft.Health.Fhir.Tests.E2E/Microsoft.Health.Fhir.Tests.E2E.csproj [hunk 17]
```diff
@@ -45,6 +45,10 @@
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
   </ItemGroup>
+
+  <ItemGroup>
```
Comment (brandonpollett) [inline test/Microsoft.Health.Fhir.Tests.E2E/Microsoft.Health.Fhir.Tests.E2E.csproj]: Was this change intentional?

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 18]
```diff
@@ -31,13 +33,15 @@ public class ControlPlaneDataStore : IControlPlaneDataStore
...
+        private readonly string _collectionId;
+        private readonly string _databaseId;
 
         public ControlPlaneDataStore(
-            IScoped<IDocumentClient> documentClient,
-            CosmosDataStoreConfiguration cosmosDataStoreConfiguration,
-            ICosmosDocumentQueryFactory cosmosDocumentQueryFactory,
-            IOptionsMonitor<CosmosCollectionConfiguration> namedCosmosCollectionConfigurationAccessor,
-            ILogger<ControlPlaneDataStore> logger)
+             IScoped<IDocumentClient> documentClient,
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: nit: extra spacing was added

File: src/Microsoft.Health.CosmosDb/Configs/CosmosDataStoreConfiguration.cs (hunk 15 above)
Comment (poadhika) [reply src/Microsoft.Health.CosmosDb/Configs/CosmosDataStoreConfiguration.cs]: Done

File: src/Microsoft.Health.ControlPlane.CosmosDb/Microsoft.Health.ControlPlane.CosmosDb.csproj (hunk 16 above)
Comment (poadhika) [reply src/Microsoft.Health.ControlPlane.CosmosDb/Microsoft.Health.ControlPlane.CosmosDb.csproj]: Not sure how these got added. Will remove it.

File: test/Microsoft.Health.Fhir.Tests.E2E/Microsoft.Health.Fhir.Tests.E2E.csproj (hunk 17 above)
Comment (poadhika) [reply test/Microsoft.Health.Fhir.Tests.E2E/Microsoft.Health.Fhir.Tests.E2E.csproj]: No and removing it.

File: src/Microsoft.Health.ControlPlane.Core/Features/Rbac/RbacService.cs [hunk 19]
```diff
@@ -23,16 +25,35 @@ public RbacService(IControlPlaneDataStore controlPlaneDataStore)
...
+
+        public async Task<IEnumerable<Role>> GetRoleForAllAsync(CancellationToken cancellationToken)
+        {
+            return await _controlPlaneDataStore.GetRoleAllAsync(cancellationToken);
+        }
+
+        public async Task<Role> UpsertRoleAsync(Role role, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(role, nameof(role));
+            return await _controlPlaneDataStore.UpsertRoleAsync(role, cancellationToken);
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.Core/Features/Rbac/RbacService.cs]: We should validate the role here. There's already a `Validate()` method on the existing `Role.cs` in `Fhir.Core.Features.Security.Role` that we can use/refactor. 

While we're at it, we should probably remove the existing one and refactor to use this. Same goes with the `ResourcePermission.cs` etc.

File: src/Microsoft.Health.ControlPlane.Core/Features/Rbac/RbacService.cs [hunk 20]
```diff
@@ -23,16 +25,35 @@ public RbacService(IControlPlaneDataStore controlPlaneDataStore)
...
+        }
+
+        public async Task<string> DeleteRoleAsync(string name, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(name, nameof(name));
+            return await _controlPlaneDataStore.DeleteRoleAsync(name, cancellationToken);
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.Core/Features/Rbac/RbacService.cs]: Similarly here we should check that this isn't the last role before removing it. If it is the last role, we should error out and not allow it. This may be easiest handled by a stored proc because we'll also need to clean up role memberships once they are implemented. Examples of stored procs are in the Fhir control plane.

File: src/Microsoft.Health.ControlPlane.Core/Features/Rbac/RbacService.cs (hunk 19 above)
Comment (poadhika) [reply src/Microsoft.Health.ControlPlane.Core/Features/Rbac/RbacService.cs]: Fixed this one

File: src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs [hunk 21]
```diff
@@ -153,5 +270,52 @@ private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitio
...
+                }
+
+                _logger.LogError(dce, "Unhandled Document Client Exception");
+                throw;
+            }
+        }
+
+        private void ValidateRole(Role role)
+        {
+            var issues = new List<OperationOutcome.IssueComponent>();
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs]: Operation outcome is a fhir specific thing and shouldn't be used here. There's a little bit of refactoring in #315 that will make this not needed.

File: src/Microsoft.Health.ControlPlane.Core/Microsoft.Health.ControlPlane.Core.csproj [hunk 22]
```diff
@@ -11,6 +11,7 @@
 
   <ItemGroup>
     <PackageReference Include="Ensure.That" Version="8.1.1" />
+    <PackageReference Include="Hl7.Fhir.STU3" Version="0.96.0" />
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.Core/Microsoft.Health.ControlPlane.Core.csproj]: Currently the control plane shouldn't have anything really fhir specific in it.

File: src/Microsoft.Health.ControlPlane.Core/Features/Rbac/Role.cs [hunk 23]
```diff
@@ -6,16 +6,29 @@
...
+            ResourcePermissions = resourcePermissions;
+            Version = version;
+        }
+
         public string Name { get; set; }
 
         public virtual string Version { get; set; }
 
-        public IList<ResourcePermission> ResourcePermissions { get; internal set; } = new List<ResourcePermission>();
+        public IList<ResourcePermission> ResourcePermissions { get; set; } = new List<ResourcePermission>();
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.Core/Features/Rbac/Role.cs]: Was this changed for a test?

File: src/Microsoft.Health.ControlPlane.Core/Features/Persistence/IControlPlaneDataStore.cs [hunk 24]
```diff
@@ -14,5 +15,15 @@ public interface IControlPlaneDataStore
...
+
+        Task<Role> GetRoleAsync(string name, CancellationToken cancellationToken);
+
+        Task<Role> UpsertRoleAsync(Role role, CancellationToken cancellationToken);
+
+        Task<IEnumerable<Role>> GetRoleAllAsync(CancellationToken cancellationToken);
+
+        Task<string> DeleteRoleAsync(string name, CancellationToken cancellationToken);
+
+        Task HardDeleteRoleAsync(string name, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (brandonpollett) [inline src/Microsoft.Health.ControlPlane.Core/Features/Persistence/IControlPlaneDataStore.cs]: We should change the name of this method to `DeleteRoleAsync` and remove the method above. Also, we shouldn't pass in a default value for cancellation token here.

File: src/Microsoft.Health.ControlPlane.Core/Features/Rbac/Role.cs (hunk 23 above)
Comment (poadhika) [reply src/Microsoft.Health.ControlPlane.Core/Features/Rbac/Role.cs]: It was. Now it has been changed.

**Comment by poadhika** [comment]
Reopening a fresh pull request
